Fix Issue 6952 - Static Linking on Linux (Add -preview=noXlinker)

JinShil opened the pull request.

JinShil committed:
Add -preview=noXlinker command-line argument

kinke commented:
I think what you are looking for is something like LDC's `-Xcc`. Those args are passed to the linker *driver* (e.g., gcc), without the `-Xlinker` prefix for linker (ld) args (`-L`).

JinShil commented:
> I think what you are looking for is something like LDC's -Xcc. Those args are passed to the linker driver (e.g., gcc), without the -Xlinker prefix for linker (ld) args (-L).

Yes, that would work.  However, I think the current behavior, where some arguments are passed to gcc and others are passed _through_ gcc, is fundamentally wrong.  What I'm trying to do with this PR is _fix_ the current behavior by adding `-preview=noXlinker` switch so we can eventually make it the default and give users a way to revert to the current behavior with a future `-revert=noXlinker` switch.

Perhaps I'm being too much of an idealist.  If this fails to achieve support, I could go with the `-Xcc` as a last resort.

kinke commented:
I don't think that breaking change would improve things, e.g.:

1. `-L--gc-sections` => `-L-Xlinker -L--gc-sections`
2. To pass `-rpath dir` to the linker: `-L-Xlinker -L-rpath -L-Xlinker -Ldir`.

With gcc's `-Wl`, the cases above could be shortened to (1) `-L-Wl,--gc-sections` and (2) `-L-Wl,-rpath,dir`, i.e., by prefixing direct linker args with IMO unpretty `-L-Wl,`. LDC went the other, backwards-compatible way of adding `-Xcc` for driver args like `-static`.

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
+            }
+        }
+
+        // Add accumulated, ordered `linkerArgs` to `argv`
```
s/Add/Append/

JinShil replied:
Addressed in https://github.com/dlang/dmd/pull/10341

JinShil commented:
Yes, this PR would make the command line more verbose for some builds, but most compiler invocations for any significant project often look like hell, and I don't think command-line beauty is all that important.  I think DMD's current unholy hybrid where some switches are passed to gcc while others are passed _through_ gcc should have never been implemented the way it is, and I'd like to undo it so we can eventually delete the code and the complexity rather than pile on yet more technical debt.

I'd be interested in hearing from others on this matter.  I'll gladly implement what the community wants, but you all have to speak up so I can be informed.

cc @ibuclaw @wilzbach @WalterBright @atilaneves @Geod24 @thewilsonator @ZombineDev  and anyone else who would like to voice their thoughts on this matter.

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
+
+                if (!isXlinker(p))
+                {
+                    if (!isNullOrEmpty(p) && !flagIsLibraryRelated(p))
```
4 instances of this logic appear. I suggest folding isNullOrEmpty into flagIsLibraryRelated.

JinShil replied:
Addressed in https://github.com/dlang/dmd/pull/10341

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
+
+        /// Dynamic array of strings to temporarily hold accumulated,
+        /// reordered linker args
+        Strings linkerArgs;
```
This is not used until line 645. Move it to where it is used.

JinShil replied:
The reason for the DDoc comment is it allows users with modern IDE to browse the code and read the comments simply by hovering one's pointer over any usage.

![image](https://user-images.githubusercontent.com/6023641/63630980-e783aa00-c65b-11e9-945a-0da9e3b947f9.png)

But I removed it in https://github.com/dlang/dmd/pull/10341 to avoid any controversy.

JinShil replied:
Addressed in https://github.com/dlang/dmd/pull/10341

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
         }
+
+        /// Dynamic array of strings to temporarily hold accumulated,
+        /// reordered linker args
```
Using Ddoc comments doesn't make sense here.

JinShil replied:
The reason for the DDoc comment is it allows users with modern IDE to browse the code and read the comments simply by hovering one's pointer over any usage.

![image](https://user-images.githubusercontent.com/6023641/63630980-e783aa00-c65b-11e9-945a-0da9e3b947f9.png)

But I removed it in https://github.com/dlang/dmd/pull/10341 to avoid any controversy.

JinShil replied:
I think I can remove `linkerArgs` altogether.

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
+            `true` if `p` starts with `needle`
+        */
+        static bool startsWith(const(char)* p, string needle)
+        {
```
Put null check here instead of at the caller.

JinShil replied:
Addressed in https://github.com/dlang/dmd/pull/10341  Please take another look.

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
+        Strings linkerArgs;
+
+        /**
+        Checks if a C string starts with a given string
```
Awkward. Better: `Checks if C string p starts with needle.`

JinShil replied:
Addressed in https://github.com/dlang/dmd/pull/10341

WalterBright commented on src/dmd/link.d:
```diff
@@ -562,59 +562,148 @@ public int runLINK()
+        Returns:
+            `true` if `p` is null or an empty string
+        */
+        static bool isNullOrEmpty(const(char)* p)
```
I don't think this function adds any value, it's only used twice.

JinShil replied:
Addressed in https://github.com/dlang/dmd/pull/10341

JinShil commented:
@WalterBright Thanks for the review.  This PR is still in draft, so I wasn't finished working on it.  Regardless, I've move the refactoring in this PR to https://github.com/dlang/dmd/pull/10341 and addressed your comments there.  Please take another look, and then let me know what you think about @kinke's comments earlier in this PR.

JinShil committed:
Add test case for -preview=noXlinker

airwin1 commented:
I have just discovered this new -preview=noXlinker command-line argument doesn't pass even a simple test to build a "hello, word" D example on Linux, see <https://issues.dlang.org/show_bug.cgi?id=20244>.  The version of dmd I tested was built from today's git master branch HEAD.  I am vitally interested in using this option so let me know if there is anyway I can help you with more testing once you feel 20244 has been fixed.

JinShil commented:
I've responded in the bug report.
